### PR TITLE
bpo-34782: Fix pdb failure when locals is an invalid mapping

### PR DESCRIPTION
**Pull request by iritkatriel**
https://bugs.python.org/issue34782

### CODE REVIEW

File: Lib/bdb.py [hunk 1]
```diff
@@ -566,10 +567,11 @@ def format_stack_entry(self, frame_lineno, lprefix=': '):
         else:
             s += "<lambda>"
         s += '()'
-        if '__return__' in frame.f_locals:
-            rv = frame.f_locals['__return__']
-            s += '->'
-            s += reprlib.repr(rv)
+        if isinstance(frame.f_locals, collections.abc.Container):
```
Comment (serhiy-storchaka) [inline Lib/bdb.py:570]: It is expected to be a mapping, not arbitrary container. For example try to pass a list.

If you want to silence a TypeError, why not use try/except?

File: Lib/pdb.py [hunk 2]
```diff
@@ -367,7 +369,8 @@ def user_return(self, frame, return_value):
         """This function is called when a return trap is set here."""
         if self._wait_for_mainpyfile:
             return
-        frame.f_locals['__return__'] = return_value
+        if isinstance(frame.f_locals, collections.abc.Mapping):
```
Comment (serhiy-storchaka) [inline Lib/pdb.py:372]: Actually it can still fail with collections.abc.Mapping. It requires MutableMapping.

But try/except will work as well.

**Review by serhiy-storchaka — COMMENTED** [review]
I am not sure there is a problem which should be fixed.

If you want to require that locals is a mapping, you should check it in `exec()` or in the frame constructor. But failing at first use can be good too.

File: Lib/bdb.py (hunk 1 above)
Comment (iritkatriel) [reply Lib/bdb.py:570]: pdb should avoid using try..except because that interferes with the exception state of the program being debugged.

Comment (serhiy-storchaka) [reply Lib/bdb.py:570]: `reprlib.repr()` and `linecache.getline()` raise and catch exceptions. And I think that the import machinery does it too.

Comment (iritkatriel) [reply Lib/bdb.py:570]: pdb does too in a few places, so we haven't reached the ideal. But do we move further away from it?

**Comment by iritkatriel** [comment]
> If you want to require that locals is a mapping, you should check it in `exec()` or in the frame constructor. But failing at first use can be good too.

If you look at the bpo, the issue here is that this code works in exec() but not in pdb(). It's not an important bug to fix, but the idea is that putting a breakpoint shouldn't make a program stop working.

**Comment by serhiy-storchaka** [comment]
It shouldn't make a _correct_ program stop working. It was incorrect program, which "worked" (actually it did not do anything useful) only by accident. Any non-trivial code will fail with invalid locals.

**Comment by iritkatriel** [comment]
> It shouldn't make a _correct_ program stop working. It was incorrect program, which "worked" (actually it did not do anything useful) only by accident. Any non-trivial code will fail with invalid locals.

That's fair. Shall we close this PR and the issue then?

**Comment by serhiy-storchaka** [comment]
I think that the only option besides just closing the issue is to add check for input arguments somewhere between frame constructor and run().

**Pull request closed without merging**